### PR TITLE
Use final Job state in undispatched coroutine completion

### PR DESCRIPTION
**Pull request by qwwdfsad**
* When using in-place coroutine builders, startUndispatchedOrReturn will return block's final value on fast-path. This is not correct because if any of the children had failed, whether or not exception is thrown depend only on child timings. Moreover, in the face of such race exception may be ignored if scope owner wasn't cancelled or is missing.
  * Introduce special hack for withTimeout and withTimeoutOrNull to ignore TimeoutCancellationException on the fast path
  * withContext, coroutineScope, supervisorScope, withTimeout and withTimeoutOrNull are affected and now check cancellation on fast-path

### CODE REVIEW

**Comment by qwwdfsad** [comment]
I am still unhappy with solution for `withTimeout` though

File: common/kotlinx-coroutines-core-common/src/intrinsics/Undispatched.kt [hunk 1]
```diff
@@ -70,29 +70,69 @@ private inline fun <T> startDirect(completion: Continuation<T>, block: () -> Any
...
+/**
+ * Same as [startUndispatchedOrReturn], but ignores [TimeoutCancellationException] on fast-path.
+ */
+internal fun <T, R> AbstractCoroutine<T>.startUndispatchedOrReturnIgnoreTimeout(
+    receiver: R, block: suspend R.() -> T,
+    timeoutCoroutine: Job
+): Any? {
+    initParentJob()
+    return undispatchedResult({ e -> !(e is TimeoutCancellationException && e.coroutine === timeoutCoroutine) })
+    { block.startCoroutineUninterceptedOrReturn(receiver, this) }
```
Comment (elizarov) [inline common/kotlinx-coroutines-core-common/src/intrinsics/Undispatched.kt]: Please, reformat to have `{` on the previous line and `}` on the next line.

File: common/kotlinx-coroutines-core-common/src/intrinsics/Undispatched.kt [hunk 2]
```diff
@@ -70,29 +70,69 @@ private inline fun <T> startDirect(completion: Continuation<T>, block: () -> Any
...
+    return undispatchedResult({ true }) { block.startCoroutineUninterceptedOrReturn(receiver, this) }
 }
 
-private inline fun <T> AbstractCoroutine<T>.undispatchedResult(startBlock: () -> Any?): Any? {
+/**
+ * Same as [startUndispatchedOrReturn], but ignores [TimeoutCancellationException] on fast-path.
+ */
+internal fun <T, R> AbstractCoroutine<T>.startUndispatchedOrReturnIgnoreTimeout(
+    receiver: R, block: suspend R.() -> T,
+    timeoutCoroutine: Job
```
Comment (qwwdfsad) [inline common/kotlinx-coroutines-core-common/src/intrinsics/Undispatched.kt]: This is redundant

File: core/kotlinx-coroutines-core/test/WithTimeoutOrNullJvmTest.kt [hunk 3]
```diff
@@ -21,4 +21,24 @@ class WithTimeoutOrNullJvmTest : TestBase() {
         // outer timeout results in null
         assertEquals(null, result)
     }
-}
\ No newline at end of file
+
+    @Test
+    fun testIgnoredTimeout() = runTest {
+        val value =withTimeout(1) {
```
Comment (qwwdfsad) [inline core/kotlinx-coroutines-core/test/WithTimeoutOrNullJvmTest.kt]: indent

File: common/kotlinx-coroutines-core-common/src/intrinsics/Undispatched.kt [hunk 4]
```diff
@@ -70,29 +70,69 @@ private inline fun <T> startDirect(completion: Continuation<T>, block: () -> Any
...
     return when {
         result === COROUTINE_SUSPENDED -> COROUTINE_SUSPENDED
         makeCompletingOnce(result, MODE_IGNORE) -> {
-            if (result is CompletedExceptionally) throw result.cause else result
+            val state = state
+            if (state is CompletedExceptionally) {
+                if (shouldThrow(state.cause)) {
+                    throw state.cause
+                } else {
+                    result
```
Comment (qwwdfsad) [inline common/kotlinx-coroutines-core-common/src/intrinsics/Undispatched.kt]: test

Comment (elizarov) [inline common/kotlinx-coroutines-core-common/src/intrinsics/Undispatched.kt]: What is result is CompletedExceptionally?

**Pull request merged**